In [ ]:
import pandas as pd
# Upload csv
#df = pd.read_csv('https://drive.google.com/uc?id=18se2kT5KanoQAhgO5Oj9d5UbB1jBJ8nX')
df = pd.read_csv('https://drive.google.com/uc?id=18se2kT5KanoQAhgO5Oj9d5UbB1jBJ8nX')
#https://drive.google.com/file/d/18se2kT5KanoQAhgO5Oj9d5UbB1jBJ8nX/view?usp=drive_link

In [ ]:
df.head()

name  year    home  \
0                   Пять ночей у Фредди (2023)  2023     США   
1                       Крэйвен-охотник (2023)  2023     США   
2                    Великий уравнитель (2014)  2014     США   
3  Трансформеры: Восхождение Звероботов (2023)  2023     США   
4                            Эскортницы (2021)  2021  Польша   

                                               genre   views  votes  \
0                         Фильмы, Ужасы, Фильмы 2023  214702    904   
1  Фильмы, Боевики, Приключения, Фантастика, Филь...  609834    666   
2   Фильмы, Боевики, Криминал, Триллеры, Фильмы 2014  457828    554   
3           Фильмы, Боевики, Приключения, Фантастика  712603   1650   
4                                      Драмы, Фильмы  583557    774   

                                             excerpt           director  \
0  Фильм «Пять ночей у Фредди» основанный на изве...        Эмма Тамми    
1  Фильм «Крэйвен-охотник» про русского иммигрант...    Джей Си Чендор    
2  «Великий уравнитель» фильм об одном спецназовц...      Антуан Фукуа    
3  В фантастическом фильме, боевике «Трансформеры...  Стивен Кейпл мл.    
4  Эми — амбициозная девушка, всегда мечтавшая о ...    Мария Садовска    

                                              actors  KP_rating  KP_votes  \
0  Джош Хатчерсон, Элизабет Лэил, Мэттью Лиллард,...        7.1    294731   
1  Аарон Тейлор-Джонсон, Рассел Кроу, Ариана ДеБо...        5.5     23497   
2  Хейли Беннетт, Билл Пуллман, Хлоя Грейс Морец,...        6.4     40699   
3  Пит Дэвидсон, Мишель Йео, Рон Перлман, Энтони ...        7.7     11061   
4  Полина Галазка, Катажина Фигура, Катажина Савч...        5.1      5026   

                                            name_eng home_eng  
0                       Five nights at Freddy (2023)      USA  
1                               Craven-hunter (2023)      USA  
2                                  Great Egor (2014)      USA  
3  Transformers: the ascent of St. John's wort (2...      USA  
4                                 Escortnitsa (2021)   Poland

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

#Choosing the features for the recommendation system

Use of better metadata and increase the quality of recomendation system.

The features taken to the recommendation are: genre, directorm actors, excerpt.

In [ ]:
features = ['genre','director', 'actors', 'excerpt']

#Cleaning the text data

Prepare the Data for getting a matrix:


*   Strip Spaces and Convert to Lowercase from all our features
*   Actors: Lesser known actors and minor roles do not really affect people's opinion of a movie. Select the major characters and their respective actors. Arbitrarily  choose the top 3 actors that appear in the list
*   Remove stop words from the excerpt
*   Remove duplicates words from combined features








In [ ]:
df.actors

0        Джош Хатчерсон, Элизабет Лэил, Мэттью Лиллард,...
1        Аарон Тейлор-Джонсон, Рассел Кроу, Ариана ДеБо...
2        Хейли Беннетт, Билл Пуллман, Хлоя Грейс Морец,...
3        Пит Дэвидсон, Мишель Йео, Рон Перлман, Энтони ...
4        Полина Галазка, Катажина Фигура, Катажина Савч...
                               ...                        
23057    Ричи Блэкмор, Грэм Боннет, Гизер Батлер, Дэвид...
23058    Рэйчел Билсон, Джейми Кинг, Кресс Уильямс, Уил...
23059    Том Эллис, Лоренц Тейт, Сара Хабель, Рик Гонса...
23060    Джастин Барта, Эндрю Рэннеллс, Джорджия Кинг, ...
23061    Максин Пик, Руперт Пенри-Джонс, Нил Стьюк, Тео...
Name: actors, Length: 23062, dtype: object

In [ ]:
df['actors'] = df['actors'].astype(str).str.lower()

In [ ]:
df['actors'] = df['actors'].astype(str).str.replace(',', '').str.split().apply(lambda x: ' '.join(x[:6]))

In [ ]:
df.actors

0           джош хатчерсон элизабет лэил мэттью лиллард
1         аарон тейлор-джонсон рассел кроу ариана дебос
2                 хейли беннетт билл пуллман хлоя грейс
3                   пит дэвидсон мишель йео рон перлман
4        полина галазка катажина фигура катажина савчук
                              ...                      
23057             ричи блэкмор грэм боннет гизер батлер
23058           рэйчел билсон джейми кинг кресс уильямс
23059                 том эллис лоренц тейт сара хабель
23060        джастин барта эндрю рэннеллс джорджия кинг
23061           максин пик руперт пенри-джонс нил стьюк
Name: actors, Length: 23062, dtype: object

In [ ]:
df['director'] = df['director'].astype(str).str.lower()

In [ ]:
df['genre'] = df['genre'].astype(str).str.lower()

In [ ]:
df['genre'].isnull().sum()

0

In [ ]:
df['excerpt'] = df['excerpt'].astype(str).str.lower()

In [ ]:
df['excerpt']

0        фильм «пять ночей у фредди» основанный на изве...
1        фильм «крэйвен-охотник» про русского иммигрант...
2        «великий уравнитель» фильм об одном спецназовц...
3        в фантастическом фильме, боевике «трансформеры...
4        эми — амбициозная девушка, всегда мечтавшая о ...
                               ...                        
23057    «тяжелый металл» называют и совестью рок-н-рол...
23058    молодой врач из нью-йорка, по имени зои харт, ...
23059    доктор уильям раш не похож на обычного терапев...
23060    сериал рассказывает о гей-паре, мечтающей имет...
23061    марта кастелло - успешный адвокат, претендующи...
Name: excerpt, Length: 23062, dtype: object

In [ ]:
# The path to stop words text file
stop_words_ru = '/content/stop.txt'


with open(stop_words_ru, 'r') as file:
    stop_words_text = file.read().splitlines()

In [ ]:
def remove_stopwords(text):
    stop_words = stop_words_text
    tokens = text.split()
    filtered_tokens = [word for word in tokens if word not in stop_words]
    return ' '.join(filtered_tokens)

In [ ]:
df['excerpt'] = df['excerpt'].apply(remove_stopwords)

In [ ]:
df['excerpt']

0        фильм «пять ночей фредди» основанный известной...
1        фильм «крэйвен-охотник» русского иммигранта им...
2        «великий уравнитель» фильм спецназовце роберте...
3        фантастическом фильме, боевике «трансформеры. ...
4        эми — амбициозная девушка, мечтавшая роскошной...
                               ...                        
23057    «тяжелый металл» называют совестью рок-н-ролла...
23058    молодой врач нью-йорка, имени зои харт, отправ...
23059    доктор уильям раш похож обычного терапевта. ра...
23060    сериал рассказывает гей-паре, мечтающей детей,...
23061    марта кастелло - успешный адвокат, претендующи...
Name: excerpt, Length: 23062, dtype: object

In [ ]:
def combine_features(row):
 return row["genre"]+" "+row["director"]+" "+row["actors"]+" "+row["excerpt"]

In [ ]:
for feature in features:
 df[feature] = df[feature].fillna('')
df["combined_features"] = df.apply(combine_features,axis=1)

In [ ]:
import re

In [ ]:
def clean_text(text):

    text = text.replace(',', '')
    # Remove numbers
    text = re.sub(r'\d+', '', text)

    # Split the text into words and remove duplicates
    words = text.split()
    words = ' '.join(sorted(set(words), key=words.index))
    return words

In [ ]:
df['combined_features'] = df['combined_features'].apply(clean_text)

In [ ]:
df["combined_features"]

0        фильмы ужасы эмма тамми джош хатчерсон элизабе...
1        фильмы боевики приключения фантастика джей си ...
2        фильмы боевики криминал триллеры антуан фукуа ...
3        фильмы боевики приключения фантастика стивен к...
4        драмы фильмы мария садовска полина галазка кат...
                               ...                        
23057    британские фильмы дик карратерс ричи блэкмор г...
23058    драмы комедии мелодрамы дэвид пэймер тим мэтис...
23059    драмы боевики деран сарафян дэвид баррет пол а...
23060    комедии мелодрамы элоди кин макс уинклер райан...
23061    драмы британские катрин морсхэд майкл оффер пи...
Name: combined_features, Length: 23062, dtype: object

In [ ]:
df.head()

name  year    home  \
0                   Пять ночей у Фредди (2023)  2023     США   
1                       Крэйвен-охотник (2023)  2023     США   
2                    Великий уравнитель (2014)  2014     США   
3  Трансформеры: Восхождение Звероботов (2023)  2023     США   
4                            Эскортницы (2021)  2021  Польша   

                                               genre   views  votes  \
0                         фильмы, ужасы, фильмы 2023  214702    904   
1  фильмы, боевики, приключения, фантастика, филь...  609834    666   
2   фильмы, боевики, криминал, триллеры, фильмы 2014  457828    554   
3           фильмы, боевики, приключения, фантастика  712603   1650   
4                                      драмы, фильмы  583557    774   

                                             excerpt           director  \
0  фильм «пять ночей фредди» основанный известной...        эмма тамми    
1  фильм «крэйвен-охотник» русского иммигранта им...    джей си чендор    
2  «великий уравнитель» фильм спецназовце роберте...      антуан фукуа    
3  фантастическом фильме, боевике «трансформеры. ...  стивен кейпл мл.    
4  эми — амбициозная девушка, мечтавшая роскошной...    мария садовска    

                                           actors  KP_rating  KP_votes  \
0     джош хатчерсон элизабет лэил мэттью лиллард        7.1    294731   
1   аарон тейлор-джонсон рассел кроу ариана дебос        5.5     23497   
2           хейли беннетт билл пуллман хлоя грейс        6.4     40699   
3             пит дэвидсон мишель йео рон перлман        7.7     11061   
4  полина галазка катажина фигура катажина савчук        5.1      5026   

                                            name_eng home_eng  \
0                       Five nights at Freddy (2023)      USA   
1                               Craven-hunter (2023)      USA   
2                                  Great Egor (2014)      USA   
3  Transformers: the ascent of St. John's wort (2...      USA   
4                                 Escortnitsa (2021)   Poland   

                                   combined_features  
0  фильмы ужасы эмма тамми джош хатчерсон элизабе...  
1  фильмы боевики приключения фантастика джей си ...  
2  фильмы боевики криминал триллеры антуан фукуа ...  
3  фильмы боевики приключения фантастика стивен к...  
4  драмы фильмы мария садовска полина галазка кат...

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df.to_csv('/content/drive/My Drive/movies_combined_features.csv', index=False)

In [ ]:
import pandas as pd
#https://drive.google.com/file/d/1LhYRK7hYfGGuGO8gQniao0Ur_HGN211L/view?usp=sharing
df = pd.read_csv('https://drive.google.com/uc?id=1LhYRK7hYfGGuGO8gQniao0Ur_HGN211L')

#Preparing the matrix of features

CV transforms the text data into a matrix of token counts. Each row in the matrix corresponds to a movie, and each column corresponds to a unique word or n-gram in the entire collection.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
# Initialize the CountVectorizer
#Replace NaN with an empty string
#df['combined_features'] = df['combined_features'].fillna('')
#cv = CountVectorizer()
cv = CountVectorizer(
    analyzer='word',        # Use word-level analysis
    ngram_range=(1, 2),     # Consider unigrams and bigrams
    min_df=0,               # Ignore terms that appear in less
    #than 0 documents (absolute counts)

)
#Replace NaN with an empty string
#df["combined_features"] = df["combined_features"].fillna('')

#Construct the required CV matrix by fitting and transforming the data
count_matrix = cv.fit_transform(df["combined_features"])


count_matrix.shape

(23062, 802556)

In the shape (23062, 802 556), the number 23062 represents the number of documents or samples in the dataset, and the number 802 556 represents the total number of unique tokens or features in the CountVectorizer matrix.

Compute the cosine similarity between each pair of documents in count_matrix. Cosine similarity is a measure of similarity between two non-zero vectors of an inner product space. It is commonly used to measure the similarity between documents based on their term frequency representations.

In [ ]:
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
cosine_sim = cosine_similarity(count_matrix,count_matrix)

In [ ]:
cosine_sim

array([[1.        , 0.03371638, 0.02301284, ..., 0.01164761, 0.01562691,
        0.        ],
       [0.03371638, 1.        , 0.04707186, ..., 0.03573708, 0.        ,
        0.        ],
       [0.02301284, 0.04707186, 1.        , ..., 0.01219603, 0.        ,
        0.        ],
       ...,
       [0.01164761, 0.03573708, 0.01219603, ..., 1.        , 0.        ,
        0.01205169],
       [0.01562691, 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.01205169, 0.        ,
        1.        ]])

#Preparing the function for getting a list of recommended movies

In [ ]:
#Construct a reverse map of indices and movie titles, that associates movie titles with their corresponding indices in the DataFrame
indices = pd.Series(df.index, index=df['name'])

In [ ]:
# Function that takes in movie title as input and outputs most similar movies with their indices and titles
def get_recommendations(name, cosine_sim=cosine_sim):
    # Get the index of the movie that matches the title
    idx = indices[name]

    # Get the pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies
    sim_scores = sim_scores[1:11]

    # Get the movie indices and similarity scores
    movie_indices = [i[0] for i in sim_scores]
    similarity_scores = [i[1] for i in sim_scores]

    #return df['name'].iloc[movie_indices]
    # Create a list of (index, title, similarity score) tuples for the top 15 most similar movies
    similar_movies = [(idx, df['name'].iloc[idx], sim_score) for idx, sim_score in zip(movie_indices, similarity_scores)]

    return similar_movies

In [ ]:
get_recommendations('Поймай меня, если сможешь (2002)')

[(14762, 'Обход Мика (2010)', 0.15839229598608032),
 (3568, 'Разыскивается (2020)', 0.12507270025494252),
 (13239, 'Военные потери (1989)', 0.12362672448829559),
 (832, 'В сердце моря (2015)', 0.12222976750248513),
 (4429, 'Секретное досье (2017)', 0.12222976750248513),
 (15660, 'Мы не рабы (2021)', 0.11506420936119476),
 (4864, 'Охотник за разумом. Схватка (2021)', 0.11238793135299599),
 (2257, 'Список Шиндлера (1993)', 0.11107897485270615),
 (19125, 'Манк (2020)', 0.10751682985805452),
 (16882, 'Поле потерянной обуви (2015)', 0.10751682985805451)]

Modify the function to get the detailed information about the recommended movies by extracting from the original df information, including the movie title in English, excerpt, genre, views, votes, and KP_rating.

In [ ]:
def improved_recommendations(name):
    # Check if the movie title exists in the dataset
    if name not in indices:
        return f"Movie '{name}' not found in the dataset."

    # Get the index and similarity scores
    idx = indices[name]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:6]
    movie_indices = [i[0] for i in sim_scores]

    # Get detailed information about the recommended movies
    recommended_movies = df.iloc[movie_indices][['name_eng', 'excerpt', 'genre', 'views', 'votes', 'KP_rating']]
    #recommended_movies = df.iloc[movie_indices][['name_eng', 'combined_features', 'genre', 'views', 'votes', 'KP_rating']]

    # Sort the recommended movies by KP_rating in descending order
    recommended_movies = recommended_movies

    # Extract the similarity scores and include them in the output
    similarity_scores = [sim_score[1] for sim_score in sim_scores]
    recommended_movies['Similarity_Score'] = similarity_scores

    return recommended_movies

In [ ]:
# Set pandas display options to show the full content of 'excerpt'
pd.set_option('display.max_colwidth', None)

In [ ]:
improved_recommendations('Матрица (1999)')

name_eng  \
2371                          Matrix: Revolution (2003)   
1803                              Matrix: reboot (2003)   
475    Transformers 3: The Dark Side of the Moon (2011)   
8050                                         Gap (2020)   
14140                             When she comes (2019)   

                                                                                                                                                                                                                                                                                                                                                                          combined_features  \
2371                    фильмы боевики фантастика лилли вачовски лана моника беллуччи лоренс фишбёрн хьюго уивинг фантастическом блокбастере «матрица : революция» кроется долгожданная развязка войны людей кибернетических организмов захвативших мир. нео товарищами оружию прошел круги ада приблизился заветной цели. играет них машины штурмуют зион оплот свободных людей. успеют...   
1803   фильмы боевики фантастика лана вачовски лилли ламбер вильсон хьюго уивинг моника беллуччи фантастический блокбастер «матрица : перезагрузка» продолжает повествовать историю борьбы людей машин поработивших человечество. нео команде удалось пережить схватку агентами матрицы задача перегруппироваться способ совершить пророчество. миссия усложнилась гонятся обоих мирах и...   
475                              фильмы боевики приключения фантастика майкл бэй джон туртурро малкович патрик демпси действия фантастической киноленты «трансформеры : темная луны» происходят спустя эпического сражения предыдущей части. сэм неоднократно спасал человечество друзьями автоботами удостоился награды рук президента. прошлом парнишка обычный офисный клерк отдыхает...   
8050           фильмы боевики фантастика джон суитц рэйчел николс каллэн мулвей angie pack центре сюжета фильма «брешь межзвездный корабль ковчег борту находится сотен человек помещенных криосон. конечная цель маршрута – новая земля остатки людей смогут восстановить цивилизацию родная планета перестала безопасным домом. экипажу состоящему профессионалов предстоит обеспечить...   
14140                                      триллеры фантастика трейлеры фильмы александр цой алевтина тукан пётр скворцов аглая тарасова офисный клерк максим свидетелем автокатастрофы. пострадавшая девушка переживает клиническую смерть длится минуты. «возвращения» максима кардинально меняется: мечты воплощаются реальность помощью таинственного дара девушки. заведут желания?...   

                                          genre  views  votes  KP_rating  \
2371                фильмы, боевики, фантастика  12315     11        7.6   
1803                фильмы, боевики, фантастика  14682      7        7.7   
475    фильмы, боевики, приключения, фантастика  53675     26        6.8   
8050   фильмы, боевики, фантастика, фильмы 2020  12772     16        5.6   
14140    триллеры, фантастика, трейлеры, фильмы     93      2        4.2   

       Similarity_Score  
2371           0.179821  
1803           0.116476  
475            0.107517  
8050           0.101149  
14140          0.098154

#Testing the recommendation system

The function improved_recommendations will be used to loop through (20% of full df) random sample of 5000 movie titles in the df and generate recommendation list of 5 movies for each movie.  

In [ ]:
# Sample 5000 movie titles from your DataFrame
movie_titles_list = df['name'].sample(n=5000, random_state=42).tolist()

In [ ]:


# Create empty lists to store the recommendations and related information
recommended_movies_list = []
views_list = []
votes_list = []
similarity_scores_list = []
mean_views_list = []
mean_votes_list = []
mean_similarity_scores_list = []

# Loop through each movie title
for movie_title in movie_titles_list:
    # Call the recommendation function for each movie title
    recommendations = improved_recommendations(movie_title)

    # Extract information from the recommendations
    recommended_movies_list.append(recommendations['name_eng'].tolist())
    views_list.append(recommendations['views'].tolist())
    votes_list.append(recommendations['votes'].tolist())

    # Calculate mean of views and votes for the recommended movies
    mean_views = recommendations['views'].mean()
    mean_votes = recommendations['votes'].mean()
    mean_views_list.append(mean_views)
    mean_votes_list.append(mean_votes)

    # Extract similarity scores from the recommendations
    similarity_scores = recommendations['Similarity_Score'].tolist()
    similarity_scores_list.append(similarity_scores)

    # Calculate mean similarity score
    mean_similarity_score = sum(similarity_scores) / len(similarity_scores)
    mean_similarity_scores_list.append(mean_similarity_score)

# Create a new DataFrame with the results
recommendations_df = pd.DataFrame({
    'Original_Movie_Title': movie_titles_list,
    'Recommended_Movies': recommended_movies_list,
    'Recommended_Views': views_list,
    'Recommended_Votes': votes_list,
    'Mean_Views': mean_views_list,
    'Mean_Votes': mean_votes_list,
    'Similarity_Scores': similarity_scores_list,
    'Mean_Similarity_Score': mean_similarity_scores_list
})



<ipython-input-40-83a08f59faeb>:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  recommended_movies['Similarity_Score'] = similarity_scores
<ipython-input-40-83a08f59faeb>:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  recommended_movies['Similarity_Score'] = similarity_scores
<ipython-input-40-83a08f59faeb>:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https:/

In [ ]:
recommendations_df

Original_Movie_Title  \
0                На месте принцессы (2018)   
1     Папа, перестань меня позорить (2021)   
2                       Боевой конь (2011)   
3                   Козёл отпущения (2012)   
4              Живущие у озера (ТВ) (1988)   
...                                    ...   
4995     Древняя любовная поэзия (1 сезон)   
4996               Бандитки в цепях (2007)   
4997                 Отряд спасения (1998)   
4998      Роман сбежавшей принцессы (2022)   
4999     Последний бросок / Лазарат (2019)   

                                                                                                                      Recommended_Movies  \
0             [Predperson (2020), Happy Bay (2020), In the place of the princess: New Life (2020), Fresno (2015), I am the groom (2018)]   
1         [Banana (2010), Some at home (2021), Christmas with losers (2004), Home Transfiguration (2010), How did we deserve it? (2018)]   
2                       [Empire of the Sun (1987), Kid (2015), Night in life (2010), Schindler list (1993), Zero bombers in fire (1984)]   
3                     [Costumer (2015), The other side of London (2018), After 2 (2020), Dirty sexual saint (2019), Days / Ri Zi (2020)]   
4                       [Room of disappointment (2016), Paranormal phenomenon 3 (2011), Patient (2022), Solitude (2015), Chamber (2010)]   
...                                                                                                                                  ...   
4995     [I am a pet in the temple Dali (season 1), Supreme Ruler (2020), Sun sisters (2021), Only next to you (2018), Soap foam (1991)]   
4996                                                 [Meow (2017), Escape (2023), Damned prison (2006), Lies (1999), Lawlessness (1989)]   
4997                    [Stranger (2010), Last rage (2017), Flashback (2022), Toronto man (2022), Detective Knight: Independence (2023)]   
4998  [Paper heart (2009), For me the last dance 2 (2006), Trip to America (1988), Pride and prejudice and dogs (2016), Rejected (2004)]   
4999    [Land of the mafia (2023), Sinjuke incident (2009), Malavita (2013), Escape (2023), Kill Bill: The whole bloody business (2011)]   

                  Recommended_Views   Recommended_Votes  Mean_Views  \
0     [4826, 1588, 5962, 780, 1509]    [26, 3, 8, 4, 4]      2933.0   
1             [67, 877, 96, 78, 77]     [0, 2, 2, 0, 2]       239.0   
2           [183, 78, 87, 1199, 86]    [2, 2, 3, 10, 2]       326.6   
3      [110, 66, 72239, 7023, 1433]  [2, 2, 263, 37, 4]     16174.2   
4          [154, 168, 180, 82, 178]     [4, 2, 2, 2, 2]       152.4   
...                             ...                 ...         ...   
4995      [595, 1420, 662, 659, 42]     [0, 3, 2, 3, 2]       675.6   
4996      [117, 891, 186, 743, 142]     [2, 3, 2, 3, 2]       415.8   
4997       [92, 93, 693, 1558, 555]     [2, 2, 4, 4, 3]       598.2   
4998         [98, 64, 505, 131, 84]     [2, 2, 4, 2, 3]       176.4   
4999    [8419, 74, 31258, 891, 459]   [14, 2, 30, 3, 3]      8220.2   

      Mean_Votes  \
0            9.0   
1            1.2   
2            3.8   
3           61.6   
4            2.4   
...          ...   
4995         2.0   
4996         2.4   
4997         3.0   
4998         2.6   
4999        10.4   

                                                                                               Similarity_Scores  \
0      [0.12051692101036454, 0.11281624057040211, 0.10482848367219184, 0.10264004785593348, 0.10129829018634484]   
1       [0.09687189593929657, 0.08481041912882635, 0.0823666225822317, 0.08225161211526649, 0.08225161211526649]   
2       [0.1399994166703125, 0.12299644696008073, 0.12010200747281088, 0.11921722634551271, 0.11097798692504908]   
3         [0.11000381964338532, 0.097212846435304, 0.09641353680829162, 0.0952448446867581, 0.09197833015863387]   
4                     [0.16, 0.15074151616205692, 0.14846149779161802, 0.14426714154678913, 0.14113006580190854]  

In [ ]:
# Calculate mean views for all recommended movies
mean_views = recommendations_df['Mean_Views'].mean()
# Calculate the mean for the 'Mean_Votes' column
mean_votes = recommendations_df['Mean_Votes'].mean()

# Calculate the mean for the 'Mean_Similarity_Score' column
mean_similarity_score = recommendations_df['Mean_Similarity_Score'].mean()

In [ ]:
mean_votes

9.375559999999998

In [ ]:
mean_similarity_score


0.1272385878614955

In [ ]:
mean_views

4362.523520000001

In [ ]:
df.to_csv('/content/drive/My Drive/test_features_CV.csv', index=False)